localstack/services/s3/provider_stream.py

```diff
@@ -0,0 +1,808 @@
+from localstack.aws.api.s3 import (
+    ChecksumAlgorithm,
+    CopyObjectOutput,
+    CopyObjectRequest,
+    InvalidStorageClass,
+    NoSuchUpload,
+    PutObjectOutput,
+    PutObjectRequest,
+    UploadPartOutput,
+    UploadPartRequest,
+)
+from localstack.services.moto import call_moto, call_moto_with_request
+from localstack.services.s3.models import get_moto_s3_backend
+from localstack.services.s3.provider import S3Provider
+from localstack.services.s3.utils import (
+    InvalidRequest,
+    extract_bucket_key_version_id_from_copy_source,
+    get_bucket_from_moto,
+    get_key_from_moto_bucket,
+    get_s3_checksum,
+    is_presigned_url_request,
+    validate_kms_key_id,
+)
+from localstack.utils.aws import arns
+from localstack.utils.patch import patch
+from localstack.utils.sync import SynchronizedDefaultDict
+
+LOG = logging.getLogger(__name__)
+
+S3_UPLOAD_PART_MIN_SIZE = 5242880  # for parity with AWS?
+S3_MAX_FILE_SIZE_BYTES = 512 * 1024
+
+MOTO_S3_DEFAULT_KEY_BUFFER_SIZE = str(S3_MAX_FILE_SIZE_BYTES)
+
+CHUNK_SIZE = 1024 * 16 * 4
+
+S3_KEY_LOCKS: dict[int, threading.RLock] = SynchronizedDefaultDict(threading.RLock)
+
+
+class S3ProviderStream(S3Provider):
+    def on_after_init(self):
+        super().on_after_init()
+        apply_stream_patches()
+
+    @handler("PutObject", expand=False)
+    def put_object(
+        self,
+        context: RequestContext,
+        request: PutObjectRequest,
+    ) -> PutObjectOutput:
+        # TODO: it seems AWS uses AES256 encryption by default now, starting January 5th 2023
+        # note: etag do not change after encryption
+        # https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-encryption.html
+
+        moto_backend = get_moto_s3_backend(context)
+        moto_bucket = get_bucket_from_moto(moto_backend, bucket=request["Bucket"])
+
+        if not config.S3_SKIP_KMS_KEY_VALIDATION and (sse_kms_key_id := request.get("SSEKMSKeyId")):
+            validate_kms_key_id(sse_kms_key_id, moto_bucket)
+
+        try:
+            request_without_body = copy.copy(request)
+            # we need to pass the query string parameters to the request to properly recreate it
+            if is_presigned_url_request(context):
+                for key, value in context.request.args.items():
+                    if key in request_without_body:
+                        request_without_body[key] = value
+                    elif key.startswith("x-amz-meta"):
+                        metadata_key = key.removeprefix("x-amz-meta-")
+                        request_without_body["Metadata"][metadata_key] = value
+
+            body = request_without_body.pop("Body", BytesIO(b""))
+            request_without_body["Body"] = BytesIO(b"")
+            checksums_keys = {
+                key for key in request_without_body.keys() if key.startswith("Checksum")
+            }
+            for checksum_key in checksums_keys:
+                request_without_body.pop(checksum_key)
```

Comment:
How do you mean by "be the same every time"?
I need to remove it because otherwise it will fail in moto. I'm passing a checksum calculated for the streaming body, but I'm passing an empty body instead, which won't match. This is one of the not great part of this PR that I don't like at all.
I'll however give it a try again without it, as some later changes might have fixed it.

Reply:
Sorry, that wasn't a proper sentence 😄 
What I meant is that a "startswith" can be dangerous, you might end up removing something you don't want to remove. But on the other hand, this approach might be more resilient to changes.
Considering that this will be changed in the next iteration anyways, it's fine I guess :)

Reply:
I actually now deleted that part, as it's not needed since we're passing an empty body and moto does not raise an exception when the checksum is wrong, at least for now. We can safely remove it and tests are still passing!